### PR TITLE
ANDROID-11447 Creating AbstractMisticaComposeView and porting Carousel to Views

### PR DESCRIPTION
**Pull request by gmerinojimenez**
### :goal_net: What's the goal?
[ANDROID-11447](https://jira.tid.es/browse/ANDROID-11447)

### :construction: How do we do it?
Preparing Mistica to have ComposeViews. To demonstrate the way of working we've ported Carousel and CarouselPageIndicator to views.

I've created an abstract base class called `AbstractMisticaComposeView` that contains the logic to know which xml brand theme is being applied and configure the compose's `MisticaTheme` accordingly. 

In the strange case where we want to have a View with a custom theme (like the one from Solar360), we can do it by setting customBrand value.

Finally, `CarouselView` and `CarouselPageIndicator` are mostly self-explanatory examples of how it works. 

### ☑️ Checks
- [X] I updated the documentation (readmes, wikis, etc). If no need to update documentation explain why.
- [X] Tested with dark mode.
- [ ] Tested with API 21.

### :test_tube: How can I test this?
https://user-images.githubusercontent.com/4595241/207855191-0f446861-59e0-4c53-8b8b-d049e678052d.mp4

### CODE REVIEW

File: catalog-compose/src/main/java/com/telefonica/mistica/compose/catalog/ui/CatalogMainActivity.kt [hunk 1]
```diff
@@ -246,6 +246,7 @@ fun Catalog(
             navigation = NavigationRoutes.CALLOUTS
         ),
     )
+        .sortedBy { it.name }
```
Comment (gmerinojimenez) [inline catalog-compose/src/main/java/com/telefonica/mistica/compose/catalog/ui/CatalogMainActivity.kt:249]: unrelated, but it was becoming very messy

image

File: library/src/main/java/com/telefonica/mistica/carousel/CarouselPageIndicatorView.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import android.util.AttributeSet
+import androidx.compose.foundation.layout.padding
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.dp
+import com.telefonica.mistica.compose.carousel.CarouselPagerIndicator
+import com.telefonica.mistica.compose.carousel.CarouselState
+import com.telefonica.mistica.compose.composeview.AbstractMisticaComposeView
+
+class CarouselPageIndicatorView @JvmOverloads constructor(
```
Comment (gmerinojimenez) [inline library/src/main/java/com/telefonica/mistica/carousel/CarouselPageIndicatorView.kt:13]: The PageIndicator are the dots below the Carousel of content
image

File: library/src/main/res/values/attrs_components.xml [hunk 3]
```diff
@@ -281,4 +281,16 @@
...
+        <attr name="composeBrand" format="enum">
+            <enum name="movistar" value="0" />
+            <enum name="movistar_prominent" value="1" />
+            <enum name="o2" value="2" />
+            <enum name="o2_classic" value="3" />
+            <enum name="vivo" value="4" />
+            <enum name="telefonica" value="5" />
+            <enum name="blau" value="6" />
+        </attr>
+    </declare-styleable>
```
Comment (gmerinojimenez) [inline library/src/main/res/values/attrs_components.xml:295]: This is the trick to know which Compose Brand theme should be applied on runtime

Comment (yamal-alm) [reply library/src/main/res/values/attrs_components.xml:295]: very smart!

File: library/src/main/java/com/telefonica/mistica/carousel/CarouselView.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    attrs,
+    defStyleAttr,
+) {
+
+    private lateinit var carouselState: CarouselState
+    private var itemCount: Int = 0
+    private lateinit var body: @Composable (Int) -> Unit
+
+    fun setState(carouselState: CarouselState): CarouselView = this.apply {
+        this.carouselState = carouselState
```
Comment (yamal-alm) [inline library/src/main/java/com/telefonica/mistica/carousel/CarouselView.kt:25]: is the view being recomposed if any of this attributes is updated?

File: library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        BRAND_VALUE_O2,
+        BRAND_VALUE_O2_CLASSIC,
+        BRAND_VALUE_VIVO,
+        BRAND_VALUE_TELEFONICA,
+        BRAND_VALUE_BLAU,
+    )
+    annotation class BrandConfig
+
+    @BrandConfig
+    var brandConfig: Int = BRAND_VALUE_TELEFONICA
```
Comment (yamal-alm) [inline library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt:50]: can this be private?

**Review by yamal-alm — APPROVED** [review]
looks very good 👏

File: library/src/main/java/com/telefonica/mistica/carousel/CarouselView.kt (hunk 4 above)
Comment (gmerinojimenez) [reply library/src/main/java/com/telefonica/mistica/carousel/CarouselView.kt:25]: it does, in fact in the Carousel example there are two ComposeViews (the carousel and the page indicator) that are sharing the same carouselState and they're updating correctly

File: library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt (hunk 5 above)
Comment (gmerinojimenez) [reply library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt:50]: yep, good catch

File: library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        BRAND_VALUE_VIVO,
+        BRAND_VALUE_TELEFONICA,
+        BRAND_VALUE_BLAU,
+    )
+    annotation class BrandConfig
+
+    @BrandConfig
+    private var brandConfig: Int = BRAND_VALUE_TELEFONICA
+
+    var customBrand: Brand? = null
```
Comment (dpastor) [inline library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt:52]: I'm thinking how we could provide this from xml or configure globally in some way, so we don't need to specify this each time a view is needed when a custom brand is provided, but i'm not sure of best approach for this.

What about some kind of a extendable map provided from a mistica object that holds mapping pairs (compose brand and xml theme)? (Currently brand objects define mapping to xml theme), probably this mapping should be **bidirectionally** placed somewhere else (and removed also from brand objects) ?

In any case we can probably do this on a next iteration.

**Review by dpastor — APPROVED** [review]
👍

File: library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt (hunk 6 above)
Comment (gmerinojimenez) [reply library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt:52]: hmm keep in mind that the compose theme is already set via xml, but you're right that we need a way to do it automatically for the custom themes. I like your idea of having an object that sets it for all the views. The only think I don't understand is why is needed to keep a map?

Comment (gmerinojimenez) [reply library/src/main/java/com/telefonica/mistica/compose/composeview/AbstractMisticaComposeView.kt]: As discussed via chat, I'm adding a MisticaComposeCustomBrand to set the custom brand for all the composeViews

**Pull request merged**